pkg/kubevirt/data_volume.go

```diff
@@ -0,0 +1,146 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	cdi "kubevirt.io/containerized-data-importer/pkg/apis/core/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// ClientFactory creates a client from the kubeconfig saved in the "kubeconfig" field of the given secret.
+type ClientFactory interface {
+	// GetClient creates a client from the kubeconfig saved in the "kubeconfig" field of the given secret.
+	// It also returns the namespace of the kubeconfig's current context.
+	GetClient(kubeconfig []byte) (client.Client, string, error)
+}
+
+// ClientFactoryFunc is a function that implements ClientFactory.
+type ClientFactoryFunc func(kubeconfig []byte) (client.Client, string, error)
+
+// GetClient creates a client from the kubeconfig saved in the "kubeconfig" field of the given secret.
+// It also returns the namespace of the kubeconfig's current context.
+func (f ClientFactoryFunc) GetClient(kubeconfig []byte) (client.Client, string, error) {
+	return f(kubeconfig)
+}
+
+// DataVolumeManager manages Kubevirt DataVolume operations.
+type DataVolumeManager interface {
+	// CreateOrUpdateVolume creates a new kubevirt Data Volume from the data volume specs and in the passed namespace.
+	CreateOrUpdateVolume(ctx context.Context, kubeconfig []byte, dataVolume *cdi.DataVolume) error
+	// GetVolume fetches the specified volume by the passed name and namespace and return DataVolumeNotFoundError error in case of
+	// not found object error.
+	GetVolume(ctx context.Context, kubeconfig []byte, name string) (*cdi.DataVolume, error)
+	// ListDataVolumes lists all the Data Volumes which exists in the passed namespace.
+	ListDataVolumes(ctx context.Context, kubeconfig []byte, listOpts ...client.ListOption) (*cdi.DataVolumeList, error)
+	// DeleteVolume delete the DataVolume based on the passed name and namespace.
+	DeleteVolume(ctx context.Context, kubeconfig []byte, name string) error
+}
+
+type defaultDataVolumeManager struct {
+	client ClientFactory
+	logger logr.Logger
+}
+
+// NewDefaultDataVolumeManager creates a new default manager with a k8s controller-runtime based on the sent kubecomfig.
+func NewDefaultDataVolumeManager(client ClientFactory) (DataVolumeManager, error) {
+	return &defaultDataVolumeManager{
+		client: client,
+		logger: log.Log.WithName("kubevirt-data-volume-manager"),
+	}, nil
+}
+
+// CreateOrUpdateVolume creates a new kubevirt Data Volume from the data volume specs and in the passed namespace.
+func (d *defaultDataVolumeManager) CreateOrUpdateVolume(ctx context.Context, kubeconfig []byte, dataVolume *cdi.DataVolume) error {
+	c, namespace, err := d.client.GetClient(kubeconfig)
+	if err != nil {
+		return errors.Wrap(err, "could not create kubevirt client")
+	}
+	dataVolume.Namespace = namespace
+
+	_, err = controllerutil.CreateOrUpdate(ctx, c, dataVolume, func() error { return nil })
```

Comment:
I strongly suspect this usage of `CrerateOrUpdate` is incorrect. It will create, but never update, please see the code of `controllerutil.CreateOrUpdate`. I think you *must* mutate the object inside the `mutateFn` last argument, otherwise the `DeepEquals` comparison will always return true and no update will even happen. To achieve this, you could pass the name and spec of the data volume as separate args, set the name / namespace before calling `CreateOrUpdate`, and set the spec inside the `mutateFn`.

Reply:
I actually agree with you, but there was an issue with having the namespace inside the mutating function. I will have another look on that one. Thanks for point that out.

Reply:
Done